lips/lip-2.md

```diff
@@ -0,0 +1,126 @@
+|ChildVASP(**addr**)                                                                                                                                   |ParentVASP                  |N                                                                                                      |-             |Y                   |Y                                                                                                                                                                                         |Y         |-                                                                                                                                         |
+
+* LibraRoot - The root authority of Libra. Controlled jointly by Libra Payment Networks and the Association Council.
+* TreasuryCompliance - A Libra Payment Networks entity responsible for day-to-day treasury (e.g. minting, burning), and compliance (e.g., updating on-chain exchange rates, freezing accounts) operations.
+* Validator - The on-chain representation of a Libra Association member.
+* ValidatorOperator - An entity authorized to operate one or more validator nodes on behalf of an Association member.
+* Designated Dealer - An entity that manages fiat transfers to/from the Libra reserve.
+* ParentVASP - The primary account of a regulated custodial wallet operating on the Libra blockchain.
+* ChildVASP(**addr**) - A child account of a the ParentVASP account at **addr**
+
+
+### Move Implementation
+Roles are implemented in the [`Roles`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Roles.md#module-0x1roles) Move module. Every account contains a [`Roles::RoleId`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Roles.md#resource-roleid) resource with an integer code to identify the role. The codes are:
+
+```
+const LIBRA_ROOT_ROLE_ID: u64 = 0;
+const TREASURY_COMPLIANCE_ROLE_ID: u64 = 1;
+const DESIGNATED_DEALER_ROLE_ID: u64 = 2;
+const VALIDATOR_ROLE_ID: u64 = 3;
+const VALIDATOR_OPERATOR_ROLE_ID: u64 = 4;
+const PARENT_VASP_ROLE_ID: u64 = 5;
+const CHILD_VASP_ROLE_ID: u64 = 6;
+```
+
+## Permissions
+The current permissions in Libra are:
+
+|Permission                                                                                                                                          |Granted to                  |Unique?                                                                                                |Transferrable?|
```

Comment:
Would it be useful to have an additional column for "granted by"? There are some cases where the code explicitly checks "granters" by assertions (e.g., PreburnCurrency, RotateDualAttenstationInfo, UpdateVMConfig).

Also, would it be unnecessary to have an additional column to specify whether granting of certain permissions is (or should be) done in genesis?

Reply:
> Would it be useful to have an additional column for "granted by"? There are some cases where the code explicitly checks "granters" by assertions (e.g., PreburnCurrency, RotateDualAttenstationInfo, UpdateVMConfig).

Very good question; I'm not very happy with the way this currently works in the code. It checks "granters" because these permissions are being added at account creation time and granter == creator. What I think we should do eventually is:
1. When creating an account, the first thing we do is generate the role capability for the account to be created
2. Privilege granting code asks for the role capability of the grantee, which the granter can pass down from step 1
This way, we'll match the discipline we discussed offline: privilege granting always checks the role of the grantee.

It actually used to work this way when we had role resources, but falling back to role checks precludes this implementation strategy.

> Also, would it be unnecessary to have an additional column to specify whether granting of certain permissions is (or should be) done in genesis?
Good question; I don't think it would be harmful, but I think it might be overspecification. The way I think of privileges is that *when* they can be granted is irrelevant (and thus need not be specified); all that matters is who can get them.